src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp

```diff
@@ -393,24 +399,36 @@ ov::npuw::CompiledModel::CompiledModel(const std::shared_ptr<ov::Model>& model,
     reset_io();
 }
 
-void ov::npuw::CompiledModel::fill_weights_bank(const std::size_t idx) {
-    LOG_VERB("Filling weights bank for Subgraph[" << idx << "]...");
-    LOG_BLOCK();
+void ov::npuw::CompiledModel::finalize_weights_bank() {
+    for (std::size_t idx = 0; idx < m_compiled_submodels.size(); ++idx) {
+        auto& comp_model_desc = m_compiled_submodels[idx];
 
-    NPUW_ASSERT(m_compiled_submodels[idx].replaced_by);
+        if (!comp_model_desc.replaced_by) {
+            continue;
+        }
 
-    auto& comp_model_desc = m_compiled_submodels[idx];
+        const auto real_idx = comp_model_desc.replaced_by.value();
+        auto& func_desc = m_compiled_submodels[real_idx];
 
-    for (std::size_t cidx = 0u; cidx < comp_model_desc.closure.size(); cidx++) {
-        comp_model_desc.closure[cidx] = m_weights_bank->update(comp_model_desc.closure[cidx]);
-        if (m_cfg.get<::intel_npu::NPUW_FOLD>()) {
-            comp_model_desc.update_required[cidx] = true;
-        } else {
-            comp_model_desc.update_required[cidx] = false;
+        // FIXME: why 0th function has those empty?
+        if (real_idx == idx) {
+            m_compiled_submodels[idx].closure.resize(m_compiled_submodels[idx].transformations.size());
+            m_compiled_submodels[idx].update_required.resize(m_compiled_submodels[idx].closure.size(),
+                                                             m_cfg.get<::intel_npu::NPUW_FOLD>() ? true : false);
+            m_compiled_submodels[idx].is_remote.resize(m_compiled_submodels[idx].closure.size(), false);
         }
-    }
 
-    LOG_VERB("DONE");
+        ov::parallel_for(comp_model_desc.transformations.size(), [&](std::size_t tidx) {
+            const auto& lt = m_compiled_submodels[idx].transformations[tidx];
+
+            if (!m_weights_bank->has(lt, *func_desc.device_it)) {
+                m_weights_bank->store(lt, lt.eval(), *func_desc.device_it);
```

Comment:
Probably only `eval` should in this `parallel_for` because bank methods have a `lock_guard` inside. Yet I'm not sure how to avoid `bank->has()`